File: LeetCode/Java/src/solutions/p1466_ReorderRoutesMakeAllPathsLeadtoCityZero/P1466.java
package solutions.p1466_ReorderRoutesMakeAllPathsLeadtoCityZero;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class P1466 {
    public int minReorder(int n, int[][] connections) {
        /*
         * Solution v1: Brute Force
         *
         * Runtime: 72 ms @ (beats) 62.64%
         * Memory Usage: 69.7 MB @ (beats) 63.89%
         */
        Map<Integer, List<Integer>> roads = new HashMap<>();
        for (int i = 0; i < n; i++) {
            roads.put(i, new ArrayList<>());
        }

        for (int[] c : connections) {
            roads.get(c[0]).add(c[1]);
            roads.get(c[1]).add(-c[0]);
        }



        return helper(n, 0, roads, new boolean[n]);
    }

    public static int helper(int n, int city, Map<Integer, List<Integer>> roads, boolean[] visited) {
        int ans = 0;

        visited[city] = true;

        for (Integer dest : roads.get(city)) {
            if (!visited[Math.abs(dest)]) {
                if (dest > 0){
                    ans++;
                }
                ans += helper(n, Math.abs(dest), roads, visited);
            }
        }
        return ans;
    }
}

File: LeetCode/Python/solutions/p1466.py
from typing import List


class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        #
        # Solution v1: Brute Force
        #
        # Runtime: 1109 ms @ (beats) 99.5%
        # Memory Usage: 61.2 MB @ (beats) 60.38%
        #
        roads = [[] for _ in range(n)]

        for s, e in connections:
            roads[s].append(e)
            roads[e].append(-s)

        visited = [False] * n

        def helper(city):
            nonlocal roads, visited
            ans = 0
            visited[city] = True

            for dest in roads[city]:
                if not visited[abs(dest)]:
                    ans += helper(abs(dest)) + (1 if dest > 0 else 0)
            return ans

        return helper(0)


if __name__ == "__main__":
    solution = Solution()

    n = 6
    connections = [[0, 1], [1, 3], [2, 3], [4, 0], [4, 5]]
    print(f"return: {solution.minReorder(n, connections)}")  # 3

    n = 5
    connections = [[1, 0], [1, 2], [3, 2], [3, 4]]
    print(f"return: {solution.minReorder(n, connections)}")  # 2

    n = 3
    connections = [[1, 0], [2, 0]]
    print(f"return: {solution.minReorder(n, connections)}")  # 0

    n = 4
    connections = [[0, 1], [2, 0], [3, 2]]
    print(f"return: {solution.minReorder(n, connections)}")  # 1
